### PR TITLE
chore(storage): implement OpenWriter, refactor gRPC write

### PR DESCRIPTION
**Pull request by noahdietz**
Implements the `storageClient.OpenWriter` inteface for `gRPC` and `HTTP`. This also refactors the existing gRPC Writer integration to use the `storageClient` implementation. This also removes the GAPIC client property from the `Client` as the `storageClient` is always used for gRPC now.

### CODE REVIEW

File: storage/http_client.go [hunk 1]
```diff
@@ -841,8 +842,92 @@ func (c *httpStorageClient) NewRangeReader(ctx context.Context, params *newRange
...
+			pr.CloseWithError(err)
+			return
+		}
+		var resp *raw.Object
+		err := applyConds("NewWriter", params.attrs.Generation, params.conds, call)
+		if err == nil {
+			if s.userProject != "" {
+				call.UserProject(s.userProject)
+			}
+			setClientHeader(call.Header())
```
Comment (cojenco) [inline storage/http_client.go:898]: I think this will be removed soon. Might be good to mark a TODO once retry headers are enabled for uploads.

File: storage/grpc_client.go [hunk 2]
```diff
@@ -956,3 +1021,303 @@ func (r *gRPCReader) reopenStream() (*storagepb.ReadObjectResponse, error) {
...
+	object        string
+	attrs         *ObjectAttrs
+	conds         *Conditions
+	encryptionKey []byte
+
+	sendCRC32C bool
+
+	// The gRPC client-stream used for sending buffers.
+	//
+	// This is an experimental API and not intended for public use.
```
Comment (cojenco) [inline storage/grpc_client.go:1068]: nit: maybe move this comment above the entire type

**Review by cojenco — COMMENTED** [review]
Awesome work🤩🎉🎉 The rearranged gRPCWriter methods are looking good. Just a few questions.

File: storage/grpc_client.go (hunk 2 above)
Comment (noahdietz) [reply storage/grpc_client.go]: I instead wrote a specific comment because I wanted to keep this one on the property itself.

File: storage/http_client.go (hunk 1 above)
Comment (noahdietz) [reply storage/http_client.go:916]: Good call out. I added a TODO for @tritone :)

File: storage/grpc_client.go [hunk 3]
```diff
@@ -964,3 +1029,301 @@ func (r *gRPCReader) reopenStream() (*storagepb.ReadObjectResponse, error) {
...
+	stream storagepb.Storage_WriteObjectClient
+
+	// The Resumable Upload ID started by a gRPC-based Writer.
+	upid string
+}
+
+// startResumableUpload initializes a Resumable Upload with gRPC and sets the
+// upload ID on the Writer.
+//
+// This is an experimental API and not intended for public use.
```
Comment (tritone) [inline storage/grpc_client.go:1086]: Can probably remove the experimental tags since it would apply to everything in this file.

File: storage/grpc_client.go [hunk 4]
```diff
@@ -964,3 +1029,301 @@ func (r *gRPCReader) reopenStream() (*storagepb.ReadObjectResponse, error) {
...
+// writeObjectSpec constructs a WriteObjectSpec proto using the Writer's
+// ObjectAttrs and applies its Conditions. This is only used for gRPC.
+//
+// This is an experimental API and not intended for public use.
+func (w *gRPCWriter) writeObjectSpec() (*storagepb.WriteObjectSpec, error) {
+	// To avoid modifying the ObjectAttrs embeded in the calling writer, deref
+	// the ObjectAttrs pointer to make a copy, then assign the desired name to
+	// the attribute.
+	attrs := *w.attrs
+	attrs.Name = w.object
```
Comment (tritone) [inline storage/grpc_client.go:1290]: I'm a little confused by this. attrs.Name is already populated by NewWriter. https://github.com/googleapis/google-cloud-go/blob/main/storage/storage.go#L1144 . So why is this necessary?

**Review by tritone — COMMENTED** [review]
Generally looks really good, sorry for the delay! Couple small things.

File: storage/writer.go [hunk 5]
```diff
@@ -277,6 +264,7 @@ func (w *Writer) Write(p []byte) (n int, err error) {
 // can be retrieved by calling Attrs.
 func (w *Writer) Close() error {
 	if !w.opened {
+		// TODO(noahdietz): also open for gRPC here? Why open here at all?
```
[the remark this answers is not in the document]
Comment (noahdietz) [reply storage/writer.go:267]: Ah, thank you, makes sense. I will add that.

File: storage/client.go [hunk 6]
```diff
@@ -211,6 +213,64 @@ type userProjectOption struct {
...
+	// Optional.
+	sendCRC32C bool
+
+	// donec is the channel used by the writer routine to indicate it is done.
+	// Required.
+	donec chan struct{}
+
+	// err is the function used by the writer routine to report errors.
+	// Required.
+	err func(error)
```
[the remark this answers is not in the document]
Comment (noahdietz) [reply storage/client.go:262]: Good call. Done.

File: storage/http_client.go [hunk 7]
```diff
@@ -857,8 +858,94 @@ func (c *httpStorageClient) NewRangeReader(ctx context.Context, params *newRange
...
+		}
+		if attrs.MD5 != nil {
+			rawObj.Md5Hash = base64.StdEncoding.EncodeToString(attrs.MD5)
+		}
+		call := c.raw.Objects.Insert(params.bucket, rawObj).
+			Media(pr, mediaOpts...).
+			Projection("full").
+			Context(params.ctx).
+			Name(params.object)
+		call.ProgressUpdater(func(n, _ int64) { progress(n) })
```
[the remark this answers is not in the document]
Comment (noahdietz) [reply storage/http_client.go:895]: It was for simplification of code, so that even if `w.ProgressFunc` is `nil`, the `w.progress` wrapper of that will no-op. This means in gRPC and HTTP we don't need the same `if w.ProgressFunc != nil {` logic. Happy to revert back to conditionally calling it though.

File: storage/grpc_client.go (hunk 3 above)
Comment (noahdietz) [reply storage/grpc_client.go:1086]: Good call. Done.

File: storage/grpc_client.go (hunk 4 above)
Comment (noahdietz) [reply storage/grpc_client.go:1290]: Ah ok, good. I was also confused because the in reviewing the existing HTTP code, it instead [references the ObjectHandle.object directly](https://github.com/googleapis/google-cloud-go/blob/main/storage/writer.go#L174) instead of the attrs. I can change this and elsewhere to use expect the attrs.Name to be set.

Comment (noahdietz) [reply storage/grpc_client.go:1290]: Note: I won't change the existing http writer code in the prod code path.

File: storage/writer.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (noahdietz) [reply storage/writer.go:267]: Done.

File: storage/grpc_client.go (hunk 4 above)
Comment (tritone) [reply storage/grpc_client.go:1290]: Cool, sounds good!

File: storage/http_client.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (tritone) [reply storage/http_client.go:895]: Ah that makes sense-- happy either way.

File: storage/client.go [hunk 8]
```diff
@@ -211,6 +213,64 @@ type userProjectOption struct {
...
 
+type openWriterParams struct {
+	// ctx is the context used by the writer routine to make all network calls
+	// and to manage the writer routine.
+	// Required.
+	ctx context.Context
+
+	// chunkSize - see `Writer.ChunkSize`.
+	// Optional.
+	chunkSize int
```
[the remark this answers is not in the document]
Comment (tritone) [reply storage/client.go:224]: Yeah makes sense. I am open to ideas, but one grouping that might make sense:
1. Passthroughs of data from the user about the object & request (name, attrs, conditions, etc)
2. Passthroughs of user configs to the Writer itself (ChunkSize, etc)
3. Pointers to shared resources with the Writer (e.g. donec) & callbacks to the writer (e.g. progress)

How does this sound?

[the remark this answers is not in the document]
Comment (noahdietz) [reply storage/client.go:224]: That sounds great. And by grouping you mean just how they are listed in this type definition, right? Or did you mean something else?

[the remark this answers is not in the document]
Comment (noahdietz) [reply storage/client.go:225]: Ok, done!

File: storage/http_client.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (noahdietz) [reply storage/http_client.go:895]: Going to leave as is for now.

File: storage/client.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (tritone) [reply storage/client.go:225]: Much more legible, thanks!
